### PR TITLE
websocket: fix deadlock when enabling/disabling via gctrpc

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

This is a quick fix for issues found in PR [here](https://github.com/thrasher-corp/gocryptotrader/pull/752#pullrequestreview-732287519) and also relates somewhat to issue #737 but a more comprehensive plan/reconstruction of that package needs to be made. 

Binance:

- Add to wait group before routine scheduling  
- Force initial sync state for orderbook management items on new connection.

Websocket: 

- When we called disable we were flushing the subscription list. Opted to generate actual state from enabled pair assets on each iteration instead. 
- Reduce leak potential on pause routine when shutdown is called while this routine is sleeping. 
- w.readMessageErrors might not have a receiver when shutdown occurs. This allows for it to return successfully. 

I have yet to test all exchange websocket functionality enable disable. But I will go through them now. 

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/binance/binance_websocket.go [hunk 1]
```diff
@@ -129,7 +140,6 @@ func (b *Binance) KeepAuthKeyAlive() {
 
 // wsReadData receives and passes on websocket messages for processing
 func (b *Binance) wsReadData() {
-	b.Websocket.Wg.Add(1)
```
Comment (gloriousCode) [inline exchanges/binance/binance_websocket.go:132]: Good pickup, but this is a widescale problem 😆 👀 . Could you fix this across all exchanges?

File: exchanges/binance/binance_websocket.go [hunk 2]
```diff
@@ -102,6 +104,15 @@ func (b *Binance) setupOrderbookManager() {
...
 		}
+	} else {
+		// Change state on reconnect for initial sync.
+		for _, m1 := range b.obm.state {
+			for _, m2 := range m1 {
+				for _, update := range m2 {
+					update.initialSync = true
+				}
+			}
+		}
```
Comment (gloriousCode) [inline exchanges/binance/binance_websocket.go:115]: You may wish to do this for bittrex and bithumb too

File: exchanges/stream/websocket.go [hunk 3]
```diff
@@ -65,8 +70,8 @@ func (w *Websocket) Setup(s *WebsocketSetup) error {
 
 	w.features = s.Features
 
-	if w.features.Subscribe && s.Subscriber == nil {
-		return errors.New("features have been set yet channel subscriber is not set")
+	if s.Subscriber == nil {
+		return errSubscriberUnset
```
Comment (gloriousCode) [inline exchanges/stream/websocket.go:74]: This one is out of scope 😄 When making sure all websockets had subscribers, GateIO does not have it as a feature under `WebsocketCapabilities: protocol.Features{`. Would you be able to add it?

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Nice work Shazbert! Here's a couple of initial nits

File: exchanges/binance/binance_websocket.go (hunk 1 above)
Comment (shazbert) [reply exchanges/binance/binance_websocket.go:132]: Yessir

File: exchanges/huobi/huobi_websocket.go [hunk 4]
```diff
@@ -125,7 +127,6 @@ func (h *HUOBI) wsFunnelConnectionData(ws stream.Connection, url string) {
 
 // wsReadData receives and passes on websocket messages for processing
 func (h *HUOBI) wsReadData() {
-	h.Websocket.Wg.Add(1)
 	defer h.Websocket.Wg.Done()
```
Comment (gloriousCode) [inline exchanges/huobi/huobi_websocket.go:130]: This function has no way of exiting and will cause a wait. Please add a shutdown listener or pass the `resp` to `comms` in `wsFunnelConnectionData` when `resp.Raw == nil` before it returns. A shutdown listener sounds simpler though

File: exchanges/bithumb/bithumb_websocket.go [hunk 5]
```diff
@@ -41,24 +41,31 @@ func (b *Bithumb) WsConnect() error {
...
 // wsReadData receives and passes on websocket messages for processing
 func (b *Bithumb) wsReadData() {
-	b.Websocket.Wg.Add(1)
 	defer b.Websocket.Wg.Done()
 
 	for {
-		resp := b.Websocket.Conn.ReadMessage()
-		if resp.Raw == nil {
+		select {
+		case <-b.Websocket.ShutdownC:
```
Comment (gloriousCode) [inline exchanges/bithumb/bithumb_websocket.go:58]: I don't believe this shutdown listener is needed in this specific instance due to your other improvements. 
If the ws is disconnected: 
- readmessage gets hit
- It will return a disconnected error
- Readmessage will attempt to relay that
  - If it cannot, you've added the default case to log a warn
- It will then return the `Response{}` with a nil `.Raw`, causing the function to return and `Wg.Done()` to be hit.

File: exchanges/stream/websocket.go [hunk 6]
```diff
@@ -210,14 +218,14 @@ func (w *Websocket) Connect() error {
 		w.connectionMonitor()
 	}
 
-	// Resubscribe after re-connection
-	if len(w.subscriptions) != 0 {
-		err = w.Subscriber(w.subscriptions)
-		if err != nil {
-			return fmt.Errorf("%v %w: %v", w.exchangeName, ErrSubscriptionFailure, err)
-		}
+	subs, err := w.GenerateSubs() // regenerate state on new connection
```
Comment (gloriousCode) [inline exchanges/stream/websocket.go:221]: More of a future warning, but if ever the exchange level `SubscribeToWebsocketChannels` function gets utilised, then this change will not pick up any additional subscriptions/unsubscriptions that have occurred when reconnecting.

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes! Only the huobi nit is important

File: exchanges/bithumb/bithumb_websocket.go (hunk 5 above)
Comment (shazbert) [reply exchanges/bithumb/bithumb_websocket.go:58]: For some reason, that was not the case. And I am unable to find out why. But this fixed the issue.

File: exchanges/stream/websocket.go (hunk 6 above)
Comment (shazbert) [reply exchanges/stream/websocket.go:221]: Would you like me to include fixes for that issue in this PR?

Comment (gloriousCode) [reply exchanges/stream/websocket.go:221]: Nah, I'll just be lying in wait, ready to nit whoever uses `SubscribeToWebsocketChannels` without updating this section

File: exchanges/bithumb/bithumb_websocket.go (hunk 5 above)
Comment (gloriousCode) [reply exchanges/bithumb/bithumb_websocket.go:58]: Thankfully what I outlined above worked for me. I won't get you to remove it, but it would have meant all `wsReadData()` implementations would need a shutdown receiver

File: exchanges/huobi/huobi_websocket.go [hunk 7]
```diff
@@ -125,13 +128,16 @@ func (h *HUOBI) wsFunnelConnectionData(ws stream.Connection, url string) {
...
 	for {
-		resp := <-comms
-		err := h.wsHandleData(resp.Raw)
-		if err != nil {
-			h.Websocket.DataHandler <- err
+		select {
+		case <-h.Websocket.ShutdownC:
+			return
```
Comment (gloriousCode) [inline exchanges/huobi/huobi_websocket.go:135]: I found that cli was not reenabling because it was stuck with a comms channel that still had an entry and no way to process it. Please use the following here to avoid failures when re-enabling a websocket:
```
case <-h.Websocket.ShutdownC:
			select {
			case <-comms:
			default:
			}
			return
```
You may also wish to use `err := h.wsHandleData(resp.Raw)` when draining the channel to not have any data loss

File: exchanges/gemini/gemini_websocket.go [hunk 8]
```diff
@@ -47,6 +47,7 @@ func (g *Gemini) WsConnect() error {
 		return err
 	}
 
+	g.Websocket.Wg.Add(2)
 	go g.wsReadData()
 	go g.wsFunnelConnectionData(g.Websocket.Conn)
```
Comment (gloriousCode) [inline exchanges/gemini/gemini_websocket.go:52]: Gemini will also require a draining of the `comms` channel on shutdown

File: exchanges/bitfinex/bitfinex_websocket.go [hunk 9]
```diff
@@ -93,7 +95,6 @@ func (b *Bitfinex) wsReadData(ws stream.Connection) {
 
 // WsDataHandler handles data from wsReadData
 func (b *Bitfinex) WsDataHandler() {
-	b.Websocket.Wg.Add(1)
 	defer b.Websocket.Wg.Done()
 	for {
 		select {
```
Comment (gloriousCode) [inline exchanges/bitfinex/bitfinex_websocket.go:100]: This will also need a `comms` drainer on shutdown. Also please place `ShutdownC` as the first `case`

File: exchanges/kraken/kraken_websocket.go [hunk 10]
```diff
@@ -153,7 +154,6 @@ func (k *Kraken) wsFunnelConnectionData(ws stream.Connection, comms chan stream.
 
 // wsReadData receives and passes on websocket messages for processing
 func (k *Kraken) wsReadData(comms chan stream.Response) {
-	k.Websocket.Wg.Add(1)
 	defer k.Websocket.Wg.Done()
 
 	for {
```
Comment (gloriousCode) [inline exchanges/kraken/kraken_websocket.go:159]: Kraken will also require a `comms` drainer

File: exchanges/binance/binance_websocket.go [hunk 11]
```diff
@@ -102,6 +104,15 @@ func (b *Binance) setupOrderbookManager() {
 			// 10 workers for synchronising book
 			b.SynchroniseWebsocketOrderbook()
```
Comment (gloriousCode) [inline exchanges/binance/binance_websocket.go:105]: `SynchroniseWebsocketOrderbook()` looks like it may require a `b.obm.jobs` drainer and `ShutdownC` as the first `case`. While `b.obm.jobs` does get reset on a new connection, the channel won't be drained in the event of a websocket shutdown.

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes! I came across an issue when re-enabling connections due to a channel not having been drained, and then the `wsReadData` functions were stuck waiting on a channel with no receiver, so the defer `wg.Done()` could not be called

**Review by gloriousCode — COMMENTED** [review]
This is looking super great. Once you fix up the test I _think_ its good to go! 🎉

**Review by thrasher- — COMMENTED** [review]
Tested enabling/disabling all websocket supported exchanges and came across this issue with for ZB (known dead WSS URL):

```
$ ./gctcliwebsocket getinfo zb
{
 "exchange": "ZB",
 "supported": true,
 "enabled": true,
 "running_url": "wss://api.zb.live/websocket"
}
$ ./gctcli websocket disable zb
2021/08/30 15:46:28 rpc error: code = Unknown desc = websocket is already disabled for exchange ZB

$ ./gctcli websocket getinfo zb
{
 "exchange": "ZB",
 "supported": true,
 "enabled": true,
 "running_url": "wss://api.zb.live/websocket"
}
```

File: exchanges/exchange.go [hunk 12]
```diff
@@ -1274,6 +1310,40 @@ func (u URL) String() string {
 	}
 }
 
+// getURLTypeFromString returns URL type from the endpoint string association
+func getURLTypeFromString(ep string) (URL, error) {
+	switch ep {
+	case "RestSpotURL":
```
Comment (gloriousCode) [inline exchanges/exchange.go:1325]: `RestSpot.String()`
etc 
etc

File: exchanges/exchange_test.go [hunk 13]
```diff
@@ -2387,3 +2387,40 @@ func TestAssetWebsocketFunctionality(t *testing.T) {
...
+		{Endpoint: "EdgeCase1URL", Expected: EdgeCase1},
+		{Endpoint: "EdgeCase2URL", Expected: EdgeCase2},
+		{Endpoint: "EdgeCase3URL", Expected: EdgeCase3},
+		{Endpoint: "sillyMcSillyBilly", Expected: 0, Error: errEndpointStringNotFound},
+	}
+
+	for _, tt := range testCases {
+		tt := tt
+		t.Run(tt.Endpoint, func(t *testing.T) {
+			u, err := getURLTypeFromString(tt.Endpoint)
```
Comment (gloriousCode) [inline exchanges/exchange_test.go:2436]: t.Parallel() ?

File: exchanges/exchange_types.go [hunk 14]
```diff
@@ -235,7 +235,7 @@ type Base struct {
 
 // url lookup consts
 const (
-	RestSpot URL = iota
+	RestSpot URL = iota + 1
```
Comment (gloriousCode) [inline exchanges/exchange_types.go:238]: how about
```
const (
Invalid URL = iota
RestSpot
...
...
```

File: exchanges/exchange.go [hunk 15]
```diff
@@ -1274,6 +1310,40 @@ func (u URL) String() string {
...
+	case "ChainAnalysisURL":
+		return ChainAnalysis, nil
+	case "EdgeCase1URL":
+		return EdgeCase1, nil
+	case "EdgeCase2URL":
+		return EdgeCase2, nil
+	case "EdgeCase3URL":
+		return EdgeCase3, nil
+	default:
+		return 0, fmt.Errorf("%w for %s", errEndpointStringNotFound, ep)
```
Comment (gloriousCode) [inline exchanges/exchange.go:1352]: Then you can use `Invalid` here and its explicit

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Nice little update
```
Binance: Config is overwriting default endpoint URL values from: [https://www.binance.com] to: [https://www.LOL.com] for: [EdgeCase1URL]
```

**Review by gloriousCode — APPROVED** [review]
retACKeroo! edit: there have been unrelated additional changes! I rescind this tack!

File: exchanges/stream/websocket.go [hunk 16]
```diff
@@ -618,6 +617,16 @@ func (w *Websocket) IsTrafficMonitorRunning() bool {
 	return w.trafficMonitorRunning
 }
 
+func (w *Websocket) checkAndSetMonitorRunning() (wasRunning bool) {
```
Comment (gloriousCode) [inline exchanges/stream/websocket.go:620]: `isRunning`/`alreadyRunning`. the past tense can have multiple interpretations

File: exchanges/stream/websocket.go [hunk 17]
```diff
@@ -298,11 +296,11 @@ func (w *Websocket) dataMonitor() {
 }
 
 // connectionMonitor ensures that the WS keeps connecting
-func (w *Websocket) connectionMonitor() {
-	if w.IsConnectionMonitorRunning() {
-		return
+func (w *Websocket) connectionMonitor() (wasRunning bool) {
```
Comment (gloriousCode) [inline exchanges/stream/websocket.go:299]: I think this one should be an error if anything is to be returned. It brings up the question why is this being called when one is already running? Even if just to log the error to highlight that there is an error in our websocket process.

**Review by gloriousCode — CHANGES_REQUESTED** [review]
More changes! 🌮

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those last changes.

**Review by thrasher- — APPROVED** [review]
tACK, thanks for addressing the nits!
